### PR TITLE
Deactivating DAGs which have been removed from files

### LINKED ISSUE #11901: DAGs remain in the UI after renaming the dag_id in the same python file
**Apache Airflow version**: 1.10.12

**What happened**:

When I rename the dag_id, the new dag_id shows up in the UI but the old dag_id does not disappear and remains in the UI. 

**What you expected to happen**:

When Airflow [lists out the python files](https://github.com/apache/airflow/blob/1.10.12/airflow/utils/file.py#L105-L161) and tries to [deactivate the deleted dags during dag processing](https://github.com/apache/airflow/blob/1.10.12/airflow/utils/dag_processing.py#L951-L955), the old DAG's file location is still in the list of the alive DAG location because the new DAG is now defined in the old DAG's python file. Even when you manually set is_active to False in the metastore, the dag processing process will set it back to True.

This happens whether or not DAG serialization is enabled.

**How to reproduce it**:

I was able to reproduce it by following these steps:
1. Add a python file in the dags folder and initialize the DAG
2. Verify the dag shows up in the UI
3. Change the dag_id in the same file to something else
4. Verify the new dag shows up in the UI
5. Both the old DAG and the new DAG are visible on the UI as well as the metadata db.

```
➜  dags pwd
/Users/alan/projects/astro/dags
➜  dags ls
wow.py
```
```python
from datetime import datetime

from airflow.models import DAG


dag = DAG(
    dag_id='yay',
    schedule_interval='@once',
    start_date=datetime(2020, 1, 1),
    catchup=False
)
```

In the screenshot, both DAGs are marked as active but there only one DAG defined in `wow.py`.
Screen Shot 2020-10-27 at 6 31 40 PM

**Anything else we need to know**:

### LINKED ISSUE #17516: Removed dynamically generated DAG doesn't always remove from Airflow metadb
**Apache Airflow version**: 2.1.0

**What happened**:

We use a python script that reads JSON configs from db to dynamically generate DAGs. Sometimes, when JSON Configs is updated, we expect DAGs to be removed in Airflow.  This doesn't always happen. From my observation, one of the following happens:
1) DAG is removed because the python script no longer generates it.
2) DAG still exists but run history is empty. When trigger the DAG, first task is stuck in `queued` status indefinitely.

**What you expected to happen**:
I expect the right behavior should be 
>1) DAG is removed because the python script no longer generates it.

**How to reproduce it**:
* Create a python script in DAG folder that dynamically generates multiple DAGs
* Execute these dynamically generated DAGs a few times
* Use Airflow Variable to toggle (reduce) the # of DAGs generated
* Examine if number of dynamically generated DAGs in the web UI

### PR DESCRIPTION
**Pull request by SamWheating**
Closes: https://github.com/apache/airflow/issues/11901
Closes: https://github.com/apache/airflow/issues/17516

Ensuring that the active DAGs in the DB are all actually present in their corresponding python files by reconciling the DB state and with the contents of a given DAG file on every parse operation. 

This _should_ prevent a lot of issues encountered when writing multiple DAGs per-file, renaming DAGs or dynamically generating DAGs based on a config file read at parse-time.

This will cause a few other functional changes:

1) DAGs will disappear from the UI if an ImportError is introduced in the underlying file.
2) DAGs will no longer be re-marked as active if they are inactive / missing but the file is present.
3) DAGs could be incorrectly marked inactive if there's an unexpected parsing error (I guess this is a tradeoff between false-poitives and false-negatives)

#### Validation:

I have validated these changes in a local breeze environment with the following DAG:

```python
from airflow.models import DAG
from airflow import utils
from airflow.operators.python import PythonOperator

NUM_DAGS=1

def message():
    print('Hello, world.')

for i in range(NUM_DAGS):
    with DAG(f'dag-{i}', schedule_interval=None, start_date=utils.dates.days_ago(1)) as dag:
        task = PythonOperator(
            task_id='task',
            python_callable=message
        )
        globals()[f"dag_{i}"] = dag

```

By changing the value of `NUM_DAGS` I can quickly change the number of DAG objects present in this file. 

Before this change, decreasing the value of `NUM_DAGS` would leave a bunch of stale DAGs in the UI. These could be triggered but would then fail as the executor was not able to load the specified task from the file.

After implementing this change, stale DAGs disappear from the UI shortly after decreasing the value of `NUM_DAGS`.

#### Questions:

1. Is there a good reason for Airflow to mark inactive DAGs as active if the file still exists? I looked through the [original PR which introduced this](https://github.com/apache/airflow/pull/5743/files) but couldn't find an explanation.

2. How significant is the performance hit incurred by updating the DAG table on every parse operation?
 
3. Will this change introduce any other functional changes that I haven't mentioned above?

#### Follow up:

If this PR is merged, we should also add some documentation which discusses dynamically generated DAGs, and explains how to generate multiple DAGs in a loop as shown in the example above.

### CODE REVIEW

File: airflow/dag_processing/processor.py [hunk 1]
```diff
@@ -618,6 +619,8 @@ def process_file(
             Stats.incr('dag_file_refresh_error', 1, 1)
             return 0, 0
 
+        self._deactivate_missing_dags(session, dagbag, file_path)
```
Comment (kaxil) [inline airflow/dag_processing/processor.py:631]: We should move this line at the end of function (or at the very least after len(dagbad.dags) check so we don't call DB if there are no dags.

File: airflow/dag_processing/processor.py [hunk 2]
```diff
@@ -647,3 +650,17 @@ def process_file(
...
+            .filter(DagModel.fileloc == file_path)
+            .filter(DagModel.is_active)
+            .filter(~DagModel.dag_id.in_(dagbag.dag_ids))
+            .all()
+        ):
+            self.log.warning(
+                "Deactivating DAG ID %s since it no longer exists in file %s", dag.dag_id, file_path
+            )
+            dag.is_active = False
+            session.merge(dag)
```
Comment (uranusjr) [inline airflow/dag_processing/processor.py]: Performance here can be improved with `UPDATE dag SET dag.is_active = FALSE WHERE ...`. I’m not sure what the right SQLAlchemy syntax is for this though.

One even significant improvement would be to somehow propagate the file paths and DAG IDs out of `DagFileProcessorProcess` and batch-upgrade all the file paths, but I don’t think that’s viable without _significant_ refactoring, so let’s not do that.

File: airflow/dag_processing/processor.py [hunk 3]
```diff
@@ -647,3 +650,17 @@ def process_file(
...
 
         return len(dagbag.dags), len(dagbag.import_errors)
+
+    def _deactivate_missing_dags(self, session: Session, dagbag: DagBag, file_path: str) -> None:
+        for dag in (
+            session.query(DagModel)
+            .filter(DagModel.fileloc == file_path)
+            .filter(DagModel.is_active)
+            .filter(~DagModel.dag_id.in_(dagbag.dag_ids))
+            .all()
```
Comment (kaxil) [inline airflow/dag_processing/processor.py]: You can chain this filter -- they use logical AND

```suggestion
            session.query(DagModel)
            .filter(DagModel.fileloc == file_path, DagModel.is_active, ~DagModel.dag_id.in_(dagbag.dag_ids))
            .all()
```

File: airflow/dag_processing/processor.py (hunk 2 above)
Comment (kaxil) [reply airflow/dag_processing/processor.py]:
```suggestion
    session.query(DagModel).filter(
        DagModel.fileloc == file_path,
        DagModel.is_active,
        ~DagModel.dag_id.in_(dagbag.dag_ids)
    ).update({"is_active": False})
```

Something like this should work !

File: airflow/dag_processing/processor.py (hunk 1 above)
Comment (SamWheating) [reply airflow/dag_processing/processor.py:631]: I think that we _do_ want to call the DB if there's no DAGs in a file - if a file used to have one DAG and now has zero then we'll need to mark that DAG is inactive, correct?

Comment (kaxil) [reply airflow/dag_processing/processor.py:631]: hmm yes True

**Comment by kaxil** [comment]
Can you rebase on latest main please

**Comment by kaxil** [comment]
@SamWheating -- I just returned from holidays, can you resolve the conflicts please

**Comment by kaxil** [comment]
One of the tests is failing, can you take a look please: https://github.com/apache/airflow/pull/17121/checks?check_run_id=3284172968#step:6:10746

**Comment by SamWheating** [comment]
> One of the tests is failing, can you take a look please: [#17121 (checks)](https://github.com/apache/airflow/pull/17121/checks?check_run_id=3284172968#step:6:10746)

I rebased my branch and have been unable to recreate this failure locally - If this failure persists in CI after a rebase then I can investigate further 👀

**Comment by kaxil** [comment]
Tests are still faining

**Comment by kaxil** [comment]
@SamWheating Can you take a look at failing tests please

**Comment by SamWheating** [comment]
I'm away on vacation for the next week - I can get these tests fixed afterwards. 

Is there a deadline for the next release?

**Comment by uranusjr** [comment]
No hard deadline yet but features probably need to finish in ~2 weeks or so to be included.

File: tests/dag_processing/test_processor.py [hunk 4]
```diff
@@ -404,7 +404,7 @@ def test_process_file_should_failure_callback(self):
 
     @conf_vars({("core", "dagbag_import_error_tracebacks"): "False"})
     def test_add_unparseable_file_before_sched_start_creates_import_error(self, tmpdir):
-        unparseable_filename = tmpdir / TEMP_DAG_FILENAME
+        unparseable_filename = os.path.join(tmpdir, TEMP_DAG_FILENAME)
```
Comment (SamWheating) [inline tests/dag_processing/test_processor.py:414]: These changes are required because the previous `tmpdir / "child_path"` syntax would return a PathLib object rather than a string, and the `_process_file` function expects a string. 

Providing the wrong type was causing some of the tests to fail due SQLAlchemy errors.

`os.path.join` will always return a string and thus fixes this issue.

Comment (uranusjr) [reply tests/dag_processing/test_processor.py:414]: `tmpdir` is not a `pathlib.Path` but a `py._path.LocalPath`, which is a path object implementation in `pytest` (actually from a package `py` that backs most of `pytest`) that pre-dates the stdlib `pathlib`.

But you fix is correct—`py._path.LocalPath` implements the new [fspath protocol](https://docs.python.org/3/library/os.html#os.PathLike) introduced in Python 3.6 (PEP 519), so it can be consumed by `os.path.join` and coerced into a str.

**Comment by uranusjr** [comment]
This seems to break `test_home_importerrors_filtered_singledag_user`, I’m gussing maybe because that DAG is not disabled and thus not triggering the `ImportError`? You’ll need to investigate this and maybe fix the test (to use another DAG or something).

**Comment by SamWheating** [comment]
Yeah, this might require some more changes, here's a shortened version of the relevant code from the rendering of the home page:

```python 

		dags_query = session.query(DagModel).filter(~DagModel.is_subdag, DagModel.is_active)    # deactivated DAGs filtered out here
        import_errors = session.query(errors.ImportError).order_by(errors.ImportError.id).all()
		all_dags = dags_query
        current_dags = all_dags

        dags = (
            current_dags.order_by(DagModel.dag_id)
            .options(joinedload(DagModel.tags))
            .offset(start)
            .limit(dags_per_page)
            .all()
        )

        if import_errors:
            dag_filenames = {dag.fileloc for dag in dags}
            all_dags_readable = (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG) in user_permissions

            for import_error in import_errors:
                if all_dags_readable or import_error.filename in dag_filenames:		# dag_filenames only includes active DAGs
                    flash(
                        "Broken DAG: [{ie.filename}] {ie.stacktrace}".format(ie=import_error),
                        "dag_import_error",
                    )
```
[source](https://github.com/apache/airflow/blob/e18b6a6d19f9ea0d8fe760ba00adf38810f0e510/airflow/www/views.py#L541)

So if a user does not have access to read all DAGs, they won't see importErrors for any DAGs since a DAG which begins throwing importErrors will be marked inactive 🤔 

I guess the fix here would be to populate `dag_filenames` from a separate query which includes deactivated DAGs? Let me know if you have any suggestions.

**Comment by SamWheating** [comment]
Fixing the remaining test case will have to wait until https://github.com/apache/airflow/pull/17924 is merged.

**Comment by SamWheating** [comment]
I don't think that the remaining failure is related to my changes 🤔

File: airflow/dag_processing/processor.py [hunk 5]
```diff
@@ -645,3 +648,12 @@ def process_file(
...
         return len(dagbag.dags), len(dagbag.import_errors)
+
+    def _deactivate_missing_dags(self, session: Session, dagbag: DagBag, file_path: str) -> None:
+        deactivated = (
+            session.query(DagModel)
+            .filter(DagModel.fileloc == file_path, DagModel.is_active, ~DagModel.dag_id.in_(dagbag.dag_ids))
+            .update({DagModel.is_active: False}, synchronize_session="fetch")
+        )
+        if deactivated:
+            self.log.info("Deactivated %i DAGs which are no longer present in %s", deactivated, file_path)
```
Comment (uranusjr) [inline airflow/dag_processing/processor.py:670]: I wonder if it’d be worthwhile to log the DAG IDs here. (Is it easy though? I know in SQL you can do `UPDATE ... RETURNING` but I don’t know what dbs can handle that nor how it’s done in SQLAlchemy.

**Review by uranusjr — APPROVED** [review]

File: airflow/dag_processing/processor.py (hunk 5 above)
Comment (SamWheating) [reply airflow/dag_processing/processor.py:670]: Thats a great idea - It looks like its supported:
https://docs.sqlalchemy.org/en/14/core/dml.html#sqlalchemy.sql.expression.Update.returning

But I'm not sure if this also means that all dialects are supported.

Comment (siddharthvp) [reply airflow/dag_processing/processor.py:670]: From https://docs.sqlalchemy.org/en/14/glossary.html#term-RETURNING:
> The backends that currently support RETURNING or a similar construct are PostgreSQL, SQL Server, Oracle, and Firebird.

Comment (SamWheating) [reply airflow/dag_processing/processor.py:670]: Ah, thanks for confirming. If its not supported in sqlite and mySQL then I don't think we can rely on this feature.

We could split this into a separated query + update like:

```python
        deactivated = (
            session.query(DagModel)
            .filter(DagModel.fileloc == file_path, DagModel.is_active, ~DagModel.dag_id.in_(dagbag.dag_ids))
        )
        if deactivated:
        	deactivated_dags = [dag.dag_id for dag in deactivated]
        	deactivated.update({DagModel.is_active: False}, synchronize_session="fetch")
        	self.log.info("Deactivated missing DAGs: %s", ",".join(deactivated_dags))
```

But I think that will introduce potential race conditions and consistency issues, so I'd rather keep things as-is.

Comment (uranusjr) [reply airflow/dag_processing/processor.py:670]: Or we can only show the IDs on Postgres and MSSQL. Or do that some other time when someone complains 😛

Comment (SamWheating) [reply airflow/dag_processing/processor.py:670]: > Or we can only show the IDs on Postgres and MSSQL.

Do you think its worth doing that? I feel like its a bit of an anti-pattern to have code like this around:
```python
if conf.get('core', 'sql_alchemy_conn').startswith(('mssql', 'postgresql')):
	# run one query
else:
	# run a different query
```

So I'd be inclined to just keep it as-is (until someone complains). Thoughts?

Comment (kaxil) [reply airflow/dag_processing/processor.py:670]: We already do things like that :)

https://github.com/apache/airflow/blob/969b239113a08f4c03ab138a0ca536ddd08a9245/airflow/jobs/scheduler_job.py#L808-L821

https://github.com/apache/airflow/blob/969b239113a08f4c03ab138a0ca536ddd08a9245/airflow/jobs/scheduler_job.py#L131-L132

Comment (kaxil) [reply airflow/dag_processing/processor.py:670]: I am touching same code, to do something similar (though not same at https://github.com/apache/airflow/pull/18120)

Comment (SamWheating) [reply airflow/dag_processing/processor.py:670]: > We already do things like that :)

Neat! Thanks for the examples - I can refactor this PR tomorrow to use different queries and logging depending on the dialect

Comment (SamWheating) [reply airflow/dag_processing/processor.py:670]: > I am touching same code, to do something similar (though not same at #18120)

Interesting, I'll follow along with the conversation there as I think that there's some functional overlap with this PR, so a lot of the discussion is probably relevant here also.

There's also another PR recently open which takes a different approach to the same problem as this PR (https://github.com/apache/airflow/pull/18087) We should definitely compare the different approaches and agree upon the best way to fix this issue.

File: airflow/models/dag.py [hunk 6]
```diff
@@ -2656,11 +2655,6 @@ def deactivate_deleted_dags(cls, alive_dag_filelocs: List[str], session=None):
             if dag_model.fileloc is not None:
                 if correct_maybe_zipped(dag_model.fileloc) not in alive_dag_filelocs:
                     dag_model.is_active = False
-                else:
-                    # If is_active is set as False and the DAG File still exists
-                    # Change is_active=True
-                    if not dag_model.is_active:
-                        dag_model.is_active = True
```
Comment (kaxil) [inline airflow/models/dag.py:2826]: Why this change? what will re-activate the DAG if it is readded

Comment (SamWheating) [reply airflow/models/dag.py:2826]: > Why this change?

With this code in place, any time a DAG is no longer present in a file it will be continually reactivated, so this code needs to be removed in order to fix this issue. 

I could not find a good explanation for why this was added in the first place ([here's the PR which introduced it](https://github.com/apache/airflow/pull/5743/files)), so I think its safe to remove. 

> what will re-activate the DAG if it is readded

I believe that the next time a DAG is found in a file, it will be marked as `active` when syncing the DAGs from the file processor to the DB.  

https://github.com/apache/airflow/blob/c73004d0cd33d76b82b172078f572e8d4eecab39/airflow/models/dag.py#L2406

Comment (kaxil) [reply airflow/models/dag.py:2826]: Yup, make sense

File: airflow/dag_processing/processor.py (hunk 5 above)
Comment (kaxil) [reply airflow/dag_processing/processor.py:670]: This one is pending or else looks good and ready to be merged

**Review by kaxil — COMMENTED** [review]
There is just one unresolved comment - gentle ping @SamWheating - https://github.com/apache/airflow/pull/17121/files#r700963529

Happy to approve and merge once that is addressed

File: airflow/dag_processing/processor.py (hunk 5 above)
Comment (SamWheating) [reply airflow/dag_processing/processor.py:670]: Ah thanks for the reminder - had a busy week. I'll try to get this changed and update the tests ASAP 😄

Comment (SamWheating) [reply airflow/dag_processing/processor.py:670]: So I just went to go and make the requested changes and it doesn't look like `update().returning()` is supported in the SqlAlchemy query API but only in direct table operations.

https://docs.sqlalchemy.org/en/14/core/dml.html#sqlalchemy.sql.expression.Update.returning

I'm not an expert in SqlAlchemy, but it looks like we use the ORM-friendly `session.query().update()` methods everywhere, in which case we can't use `.returning()`

Does my understanding sound correct? If so then we'll have to just mark this one as resolved.

**Review by kaxil — APPROVED** [review]

**Comment by kaxil** [comment]
Kubernetes test failure are unrelated
